File: java/src/blog/ds/StackDynamicArray.java
package blog.ds;

import java.util.Arrays;
import java.util.stream.Collectors;

public class StackDynamicArray {
    private int top = 0;

    private int[] arr = new int[]{0};

    public String toString() {
        return Arrays.stream(arr).mapToObj(String::valueOf).collect(Collectors.joining(","));
    }

    public void push(int data) {
        if (this.top >= this.arr.length) {
            int length = this.arr.length;
            int[] newArr = new int[length * 2];

            for(int i = 0; i < length; i++)
                newArr[i] = this.arr[i];

            this.arr = newArr;
        }
        this.arr[this.top++] = data;
    }

    public Integer pop() {
        if (this.empty()) {
            return null;
        } else if (this.top < this.arr.length / 2) {
            int length = this.arr.length;
            int[] newArr = new int[length / 2];

            for(int i = 0; i < this.top; i++)
                newArr[i] = this.arr[i];

            int result = this.arr[--this.top];
            this.arr = newArr;

            return result;
        } else {
            return this.arr[--this.top];
        }
    }

    public boolean empty() {
        return this.top == 0;
    }

    public Integer peek() {
        if (empty())
            return null;
        return this.arr[this.top - 1];
    }

    public int search(int data) {
        for (int i = 0; i < this.top; i++) {
            if (data == this.arr[i])
                return i;
        }
        return -1;
    }
}

File: python/blog/ds/stackDynamicArray.py
class StackDynamicArray:
    top = 0
    arr = [-1]

    def toString(self):
        return ','.join(map(str, self.arr))

    def _createNewArray(self):
        length = len(self.arr)
        newArr = [-1 for i in range(length * 2)]

        for i in range(length):
            newArr[i] = self.arr[i]

        self.arr = newArr

    def push(self, data):
        if self.top >= len(self.arr):
            self._createNewArray()
        
        self.arr[self.top] = data
        self.top += 1

    def _shrinkArray(self):
        length = len(self.arr)
        newArr = [-1 for i in range(length // 2)]

        for i in range(self.top):
            newArr[i] = self.arr[i]

        self.top -= 1
        result = self.arr[self.top]

        self.arr = newArr

        return result

    def pop(self):
        if self.empty():
            return None
        elif self.top < len(self.arr) // 2:
            return self._shrinkArray()
        else:
            self.top -= 1
            return self.arr[self.top]

    def empty(self):
        return self.top == 0

    def peek(self):
        if self.empty():
            return None
        return self.arr[self.top - 1]

    def search(self, data):
        for i in range(self.top):
            if data == self.arr[i]:
                return i
        return -1
    
